aten/src/ATen/native/BucketizationUtils.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <ATen/ATen.h>
+
+namespace at {
+namespace native {
+
+template<typename F>
+void searchsorted_generic_template(
+  Tensor& result,
+  const Tensor& input,
+  const Tensor& boundaries,
+  bool right,
+  F&& contiguous_solver) {
+
+  bool in_is_contiguous = input.is_contiguous();
+  bool bd_is_contiguous = boundaries.is_contiguous();
+
+  if (in_is_contiguous && bd_is_contiguous) {
+    contiguous_solver(result, input, boundaries, right);
+    return;
+  }
+
+  TORCH_WARN_ONCE("contiguous input tensors are expected, but got boundaries tensor ",
+    (bd_is_contiguous ? "is contiguous" : "is not contiguous"), " and input value tensor ",
+    (in_is_contiguous ? "is contiguous" : "is not contiguous"), " the non-contiguous tensors "
+    "have been temporarily duplicated into contiguous tensors, this process has data copy between "
+    "tensors which lower the performance, please make sure the input tensors are contiguous!");
+
+  if (!in_is_contiguous && bd_is_contiguous) {
+    Tensor input_cont = input.contiguous();
+    contiguous_solver(result, input_cont, boundaries, right);
+    return;
+  }
+
+  Tensor boundaries_cont = boundaries.contiguous();
+  if (in_is_contiguous) {
+    contiguous_solver(result, input, boundaries, right);
+  }
+  else {
+    Tensor input_cont = input.contiguous();
+    contiguous_solver(result, input_cont, boundaries_cont, right);
+  }
+}
+
+
+inline bool searchsorted_dims_matched_before_last_dim(const Tensor& boundaries, const Tensor& input) {
+  if (boundaries.dim() != input.dim()) {
+    return false;
+  }
+
+  const auto& dims_bd = boundaries.sizes();
+  const auto& dims_in = input.sizes();
+  for (int64_t dim = 0; dim + 1 < boundaries.dim(); ++dim) {
+    if (dims_bd[dim] != dims_in[dim]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+inline void searchsorted_pre_check(const Tensor& boundaries, const Tensor& input, const Tensor& output, bool out_int32) {
+  TORCH_CHECK(boundaries.device() == input.device(), "boundaries and input value tensors should have same device type, ",
+    "but we got boundaries tensor device type ", boundaries.device(), " and input value tensor device type ", input.device());
+
+  TORCH_CHECK(boundaries.dtype() == input.dtype(), "boundaries and input value tensors should have same dtype, ",
```

Comment:
do we want this function to not type-promote? numpy type-promotes.

Reply:
@ngimel @VitalyFedyunin @gchanan Do we have a standard way (or an existed api) to determine what type should be promoted to what type? or I need to defined the rule here? or I can just promoted them all to double if dtype are not match between tensors?

Reply:
unfortunately, we don't have free-standing usable type promotion API, TensorIterator is doing some promotion, but it's spread around several functions. Definitely don't promote everything to double. 
cc @mruberry - what type promotions rules should we follow here, if any?

Reply:
Ideally we'd implement in-kernel comparators for the different scalar types, but that seems like scope creep for this PR. It's OK to require the input tensors share a dtype for now.

Reply:
found one here. Should be good to use.
https://github.com/pytorch/pytorch/blob/dece155335eaf808df54474fd4efaa65b694fde8/c10/core/ScalarType.h#L388